File: ios/Kwift/com/lightningkite/kwift/views/ios/SelectDateRangeMonthView.swift
//
//  SelectDayMonthView.swift
//  KwiftTemplate
//
//  Created by Joseph Ivie on 9/11/19.
//  Copyright © 2019 Joseph Ivie. All rights reserved.
//

import Foundation
import UIKit


public class SelectDateRangeMonthView : QuickMonthView {
    var start: MutableObservableProperty<Date?> = StandardObservableProperty(nil) {
        didSet {
            startup2()
            setNeedsDisplay()
        }
    }
    var endInclusive: MutableObservableProperty<Date?> = StandardObservableProperty(nil) {
        didSet {
            startup2()
            setNeedsDisplay()
        }
    }
    public override init(frame: CGRect) {
        super.init(frame: frame)
        startup2()
    }

    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        startup2()
    }

    var closers: Array<Close> = []
    public func startup2() {
        for close in closers {
            close.close()
        }
        closers.append(start.onChange.addWeak(self) { this, value in
            this.setNeedsDisplay()
        })
        closers.append(endInclusive.onChange.addWeak(self) { this, value in
            this.setNeedsDisplay()
        })
    }

    override public func drawDayCell(_ ctx: CGContext, _ rect: CGRect, _ day: Date) {
        guard let s = start.value, let e = endInclusive.value else {
            if let s = start.value, s.sameDay(day) {
                drawDayCellBackground(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
                drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            } else {
                drawDayCellBackground(ctx: ctx, rectangle: rect, day: day, colorSet: self.defaultColorSet)
                drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.defaultColorSet)
            }
            return
        }

        if s.sameDay(e), day.sameDay(s) {
            drawDayCellBackground(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
        } else if s.sameDay(day) {
            drawDayCellBackgroundStart(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
        } else if e.sameDay(day) {
            drawDayCellBackgroundEnd(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
        } else if day.after(s), day.before(e) {
            drawDayCellBackgroundMiddle(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
        } else {
            drawDayCellBackground(ctx: ctx, rectangle: rect, day: day, colorSet: self.defaultColorSet)
            drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.defaultColorSet)
        }
    }

    private var startedDraggingOn: Date = Date()
    private var everMoved = false
    var draggingStart = true
    
    override public func onTouchDown(date: Date) -> Bool {
        startedDraggingOn = date
        everMoved = false
        
        guard let startValue = start.value, let endInclusiveValue = endInclusive.value else {
            start.value = date
            endInclusive.value = date
            return true
        }
        if date.sameDay(startValue) {
            draggingStart = true
        } else if date.sameDay(endInclusiveValue){
            draggingStart = false
        } else if date.after(endInclusiveValue), startValue.sameDay(endInclusiveValue) {
            endInclusive.value = date
            draggingStart = false
        } else {
            start.value = date
            endInclusive.value = date
            draggingStart = false
        }
        return true
    }
    override public func onTouchMove(date: Date) -> Bool {
        if !date.sameDay(startedDraggingOn) {
            everMoved = true
        }
        if let startValue = start.value, let endInclusiveValue = endInclusive.value {
            if draggingStart, date.after(endInclusiveValue) {
                start.value = endInclusiveValue
                endInclusive.value = date
                draggingStart = false
            } else if !draggingStart, date.before(startValue) {
                endInclusive.value = startValue
                start.value = date
                draggingStart = true
            }
        }

        if draggingStart {
            start.value = date
        } else {
            endInclusive.value = date
        }
        return true
    }
    
    override public func onTouchUp(date: Date) -> Bool {
        let result = onTouchMove(date: date)
        if startedDraggingOn.sameDay(date) && !everMoved {
            start.value = date
            endInclusive.value = date
        }
        return result
    }
}

File: android/src/main/java/com/lightningkite/kwift/views/android/SelectDateRangeMonthView.kt
package com.lightningkite.kwift.views.android

import android.content.Context
import android.graphics.Canvas
import android.graphics.RectF
import android.util.AttributeSet
import com.lightningkite.kwift.observables.Close
import com.lightningkite.kwift.observables.MutableObservableProperty
import com.lightningkite.kwift.observables.StandardObservableProperty
import com.lightningkite.kwift.observables.addWeak
import java.util.*

open class SelectDateRangeMonthView : QuickMonthView {

    var draggingStart = true
    var start: MutableObservableProperty<Calendar?> = StandardObservableProperty(null)
        set(value) {
            field = value
            init()
            invalidate()
        }
    var endInclusive: MutableObservableProperty<Calendar?> = StandardObservableProperty(null)
        set(value) {
            field = value
            init()
            invalidate()
        }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    val closers: ArrayList<Close> = ArrayList()
    fun init() {
        for (close in closers) {
            close.close()
        }
        closers += start.onChange.addWeak(this) { self, value ->
            self.invalidate()
        }
        closers += endInclusive.onChange.addWeak(this) { self, value ->
            self.invalidate()
        }
    }

    override fun drawDayCell(canvas: Canvas, rectangle: RectF, day: Calendar) {
        when {
            day sameDay start.value && (day sameDay endInclusive.value || endInclusive.value == null) -> {
                drawDayCellBackground(canvas, rectangle, selectedColorSet)
                drawDayCellText(day, canvas, rectangle, selectedColorSet)
            }
            day sameDay start.value -> {
                drawDayCellBackgroundStart(canvas, rectangle, selectedColorSet)
                drawDayCellText(day, canvas, rectangle, selectedColorSet)
            }
            day sameDay endInclusive.value -> {
                drawDayCellBackgroundEnd(canvas, rectangle, selectedColorSet)
                drawDayCellText(day, canvas, rectangle, selectedColorSet)
            }
            day.timeInMillis > start.value?.timeInMillis ?: Long.MAX_VALUE && day.timeInMillis < endInclusive.value?.timeInMillis ?: Long.MIN_VALUE -> {
                drawDayCellBackgroundMiddle(canvas, rectangle, selectedColorSet)
                drawDayCellText(day, canvas, rectangle, selectedColorSet)
            }
            else -> {
                drawDayCellBackground(canvas, rectangle, defaultColorSet)
                drawDayCellText(day, canvas, rectangle, defaultColorSet)
            }
        }
    }

    private var startedDraggingOn: Calendar? = null
    private var everMoved = false
    private var nonModifyingSoFar = false

    override fun onTouchDown(calendar: Calendar): Boolean {
        startedDraggingOn = calendar
        everMoved = false
        nonModifyingSoFar = true
        //If on start/end - drag
        //If after, extend
        //If before, extend
        //If middle, collapse all
        val startValue = start.value
        val endInclusiveValue = endInclusive.value

        when {
            startValue == null || endInclusiveValue == null -> {
                start.value = calendar
                endInclusive.value = calendar
                draggingStart = false
                nonModifyingSoFar = false
            }
            calendar sameDay endInclusiveValue -> {
                draggingStart = false
            }
            calendar sameDay startValue -> {
                draggingStart = true
            }
            calendar.after(endInclusiveValue) && startValue sameDay endInclusiveValue -> {
                endInclusive.value = calendar
                draggingStart = false
                nonModifyingSoFar = false
            }
            else -> {
                start.value = calendar
                endInclusive.value = calendar
                draggingStart = false
                nonModifyingSoFar = false
            }
        }
        return true
    }

    override fun onTouchMove(calendar: Calendar): Boolean {

        if(!(calendar sameDay startedDraggingOn)){
            everMoved = true
        }
        val startValue = start.value
        val endInclusiveValue = endInclusive.value
        when {
            startValue == null || endInclusiveValue == null -> {
            }
            draggingStart && calendar.after(endInclusiveValue) -> {
                start.value = endInclusive.value
                endInclusive.value = calendar
                nonModifyingSoFar = false
                draggingStart = false
                return true
            }
            !draggingStart && calendar.before(startValue) -> {
                endInclusive.value = start.value
                start.value = calendar
                nonModifyingSoFar = false
                draggingStart = true
                return true
            }
        }

        val obs = if (draggingStart) start else endInclusive
        if(obs.value != calendar){
            nonModifyingSoFar = false
        }
        obs.value = calendar
        return true
    }

    override fun onTouchUp(calendar: Calendar): Boolean {
        onTouchMove(calendar)
        if(startedDraggingOn sameDay calendar && !everMoved && nonModifyingSoFar){
            start.value = calendar
            endInclusive.value = calendar
        }
        return true
    }

    private infix fun Calendar?.sameDay(other: Calendar?): Boolean =
        this != null && other != null &&
                this.get(Calendar.YEAR) == other.get(Calendar.YEAR) &&
                this.get(Calendar.MONTH) == other.get(Calendar.MONTH) &&
                this.get(Calendar.DAY_OF_MONTH) == other.get(Calendar.DAY_OF_MONTH)
}
